osutil/mkdir.go

```diff
@@ -0,0 +1,164 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2024 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package osutil
+
+import (
+	"os"
+	"path/filepath"
+	"sync"
+	"syscall"
+
+	"github.com/snapcore/snapd/osutil/sys"
+)
+
+// XXX: we need to come back and fix this; this is a hack to unblock us.
+// Have a lock so that if one goroutine tries to Mkdir /foo/bar, and
+// another tries to Mkdir /foo/baz, they can't both decide they need
+// to make /foo and then have one fail.
+var mu sync.Mutex
```

Comment:
instead of forcing synchronization, maybe simply document that for the caller? although as a consumer of the API I would not expect anything different than the racy behavior, which I think is consistent with how things generally work across different languages/libraries.

And in fact you cannot do anything about racing with other processes, so I'm not sure it's worth to even worry about this.

Reply:
Hi, thanks for pointing it out, but this logic is in the current code base, see [here](https://github.com/canonical/snapd/blob/master/osutil/mkdirallchown.go#L35).

This PR only refactors the API interface and does not add new or change existing logic. The API is renamed, and so is the file name, making it hard to show a diff for reviewing.